### PR TITLE
[react-redux] Fix map factories

### PR DESCRIPTION
**Pull request by NicholasBoll**
* Changed the order of factory vs object in mapping functions so TS picks the right one
* Added test cases and test style

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Increase the version number in the header if appropriate.

### CODE REVIEW

File: types/react-redux/index.d.ts [hunk 1]
```diff
@@ -141,7 +142,7 @@ interface MapStateToPropsFactory<TStateProps, TOwnProps> {
     (initialState: any, ownProps: TOwnProps): MapStateToProps<TStateProps, TOwnProps>;
 }
 
-type MapStateToPropsParam<TStateProps, TOwnProps> = MapStateToProps<TStateProps, TOwnProps> | MapStateToPropsFactory<TStateProps, TOwnProps>;
+type MapStateToPropsParam<TStateProps, TOwnProps> = MapStateToPropsFactory<TStateProps, TOwnProps> | MapStateToProps<TStateProps, TOwnProps> | null | undefined;
```
Comment (NicholasBoll) [inline types/react-redux/index.d.ts]: The previous order caused a map factory to not be inferred. Listing the factory first forces TS to recognized the factory first and thus infer `TStateProps` properly

File: types/react-redux/react-redux-tests.tsx [hunk 2]
```diff
@@ -10,6 +10,265 @@ import objectAssign = require('object-assign');
...
+    })
+    
+    const mapDispatchToProps = () => ({
+        onClick: () => {}
+    })
+
+    class TestComponent extends Component<OwnProps & StateProps & DispatchProps> {}
+
+    const Test = connect(
+        mapStateToPropsFactory,
```
Comment (NicholasBoll) [inline types/react-redux/react-redux-tests.tsx]: This had a TS error previously. `{ bar: number }` was not inferred as a `TStateProps` and thus became required in the verify section

File: types/react-redux/react-redux-tests.tsx [hunk 3]
```diff
@@ -10,6 +10,265 @@ import objectAssign = require('object-assign');
...
+        mapDispatchToProps,
+    )(TestComponent)
+
+    const verify = <Test foo='bar' />
```
Comment (NicholasBoll) [inline types/react-redux/react-redux-tests.tsx]: I find this style of verification steps more organized and easier to understand what is being tested. Many existing type tests don't verify the output of `connect` and many breakages and fixes have occurred. The `verify` is like an expectation or an assertion. It is a way to verify the component returned by `connect` has the right property requirements in a way that is understandable by people just doing React.

This line could be removed and the test look like this instead:

```ts
const Test: React.ComponentClass<{ foo: string }> = connect(
  mapStateToPropsFactory,
  mapDispatchToProps,
)(TestComponent)
```

But this method allows cast narrowing. Meaning `React.ComponentClass<{}>` would be a valid cast target (`{}` can fit into `{ foo: string }`, but not the reverse)

**Comment by NicholasBoll** [comment]
I experimented with rewriting the type definitions and was able to pass every existing test with 5 less overrides, but I didn't want to muddy the waters for this 1 defect. Perhaps that is something I could pursue if the maintainers would be okay with it.

File: types/react-redux/react-redux-tests.tsx [hunk 4]
```diff
@@ -10,6 +10,265 @@ import objectAssign = require('object-assign');
 // https://github.com/rackt/react-redux/blob/master/docs/quick-start.md#quick-start
 //
 
+// Test cases written in a way to isolate types and variables and verify the
+// output of `connect` to make sure the signature is what is expected
+
+namespace Empty {
```
Comment (NicholasBoll) [inline types/react-redux/react-redux-tests.tsx]: namespacing allows all type and variable definitions to be isolated preventing a lot of numbers from being appended to types and variables making the test easier to understand and isolated. It was a bit odd that if I forgot to defined `mapStateToProps` that I'd get a type error about not having `{ value: number }` since that is the signature of a globally defined function

File: types/react-redux/tsconfig.json [hunk 5]
```diff
@@ -22,4 +22,4 @@
         "noEmit": true,
         "forceConsistentCasingInFileNames": true
     }
-}
```
Comment (NicholasBoll) [inline types/react-redux/tsconfig.json]: I noticed some types were off in the test file and originally changed `strictNullChecks` to true, but I didn't know if there was a backstory to why this was set to false. I think `strictNullChecks` should always be set to true, otherwise TS is perfectly happy with the following:

```ts
function getLength(value: string): number {
  return value.length;
}

getLength(null); // TS is happy with `strictNullChecks=false`, but this will result in a runtime TypeError
```

Comment (clayne11) [reply types/react-redux/tsconfig.json]: I'm not sure if there's a backstory either but I definitely agree with you on this one.

**Review by clayne11 — APPROVED** [review]

File: types/react-redux/tsconfig.json (hunk 5 above)
Comment (thasner) [reply types/react-redux/tsconfig.json]: same

File: types/react-redux/index.d.ts (hunk 1 above)
Comment (thasner) [reply types/react-redux/index.d.ts]: late to the 🎉 :
was adding `| null | undefined` necessary? any reason you didn't do that for `MapDispatchToPropsParam`?

Comment (clayne11) [reply types/react-redux/index.d.ts]: It's necessary because often times you only want to give an argument for dispatch without requiring state which means you need to pass null for `mapStateToProps`.

Comment (thasner) [reply types/react-redux/index.d.ts]: wouldn't the same be true for only providing an argument for `mergeProps`?

```
export const FooHOC = connect<undefined, undefined, TOwnProps, TMergedProps>(
    undefined,
    undefined,
    mergeProps,
)(UnconnectedFoo);
```

I have similar code in a codebase with `strictNullChecks`, and it works fine with current typings.

Comment (thasner) [reply types/react-redux/index.d.ts]: on second thought, this goes against the overloads we've provided.
the idea is that overloads that expect `mapStateToProps` to be defined reference this type, and those who don't expect `null | undefined` explicitly.

e.g.
https://github.com/NicholasBoll/DefinitelyTyped/blob/1cc622c196e2451bd2272e2af87221fc82da2dec/types/react-redux/index.d.ts#L76
vs
https://github.com/NicholasBoll/DefinitelyTyped/blob/1cc622c196e2451bd2272e2af87221fc82da2dec/types/react-redux/index.d.ts#L81

Comment (clayne11) [reply types/react-redux/index.d.ts]: Good catch. I don't think the changes are going to cause any issues but we should probably change this one piece back to the way that it was since it's unnecessary.

Comment (NicholasBoll) [reply types/react-redux/index.d.ts]: Good catch. Originally I was able to get right of 4 extra overloads, but decided against the more drastic change. The `null | undefined` was leftover from that revert. It doesn't actually cause harm in this case, but currently the overloads don't work as expected anyways. Oftentimes the unexpected overload is picked (`connect(mapStateToProps, mapStateToDispatch)(Component)` will pick an overload with `mergeProps`. I was also experimenting forcing some overloads to extend from `{}` or `object` to force Typescript to recognize the correct overload.

I thought we should add more test cases before refactoring anything though. It should be safe to remove `null | undefined` as that edge case shouldn't be hit with the current typings.
